x-pack/filebeat/input/azureeventhub/v2_input.go

```diff
@@ -0,0 +1,532 @@
+				Cloud: cloud.AzurePublic,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create blob container pipelineClient: %w", err)
+	}
+
+	// The modern event hub SDK does not create the container
+	// automatically like the old SDK.
+	//
+	// The new `BlobStore` explicitly says:
+	//   "the container must exist before the checkpoint store can be used."
+	//
+	// We need to ensure it exists before we can use it.
+	err = in.ensureContainerExists(ctx, containerClient)
+	if err != nil {
+		return fmt.Errorf("failed to ensure blob container exists: %w", err)
+	}
+
+	// The checkpoint store is used to store the checkpoint information
+	// in the blob container.
+	checkpointStore, err := checkpoints.NewBlobStore(containerClient, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create checkpoint store: %w", err)
+	}
+	in.checkpointStore = checkpointStore
+
+	// Create the event hub consumerClient to receive events.
+	consumerClient, err := azeventhubs.NewConsumerClientFromConnectionString(
+		in.config.ConnectionString,
+		in.config.EventHubName,
+		in.config.ConsumerGroup,
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create consumer pipelineClient: %w", err)
+	}
+	in.consumerClient = consumerClient
+
+	// Manage the migration of the checkpoint information
+	// from the old Event Hub SDK to the new Event Hub SDK.
+	in.migrationAssistant = newMigrationAssistant(
+		in.log,
+		consumerClient,
+		containerClient,
+		checkpointStore,
+	)
+
+	return nil
+}
+
+// run starts the main loop for processing events.
+func (in *eventHubInputV2) run(ctx context.Context) {
+	if in.config.MigrateCheckpoint {
+		in.log.Infow("checkpoint migration is enabled")
+		// Check if we need to migrate the checkpoint store.
+		err := in.migrationAssistant.checkAndMigrate(
+			ctx,
+			in.config.ConnectionString,
+			in.config.EventHubName,
+			in.config.ConsumerGroup,
+		)
+		if err != nil {
+			in.log.Errorw("error migrating checkpoint store", "error", err)
+			// FIXME: should we return here?
+		}
+	}
+
+	// Handle the case when the processor stops due to
+	// transient errors (network failures) and we need to
+	// restart it.
+	processorRunBackoff := backoff.NewEqualJitterBackoff(
+		ctx.Done(),
+		processorRestartBackoff,    // initial backoff
+		processorRestartMaxBackoff, // max backoff
+	)
+
+	// Create the processor options using the input
+	// configuration.
+	processorOptions := createProcessorOptions(in.config)
+
+	for ctx.Err() == nil {
+		// Create a new processor for each run.
+		//
+		// The docs explicitly say that the processor
+		// is not reusable.
+		processor, err := azeventhubs.NewProcessor(
+			in.consumerClient,
+			in.checkpointStore,
+			processorOptions,
+		)
+		if err != nil {
+			in.log.Errorw("error creating processor", "error", err)
+			return
+		}
+
+		// Launch one goroutines for each partition
+		// to process events.
+		go in.workersLoop(ctx, processor)
+
+		// Run the processor to start processing events.
```

Comment:
https://github.com/Azure/azure-sdk-for-go/blob/c2697b22b5508e23768fc943419b109110677128/sdk/messaging/azeventhubs/example_consuming_with_checkpoints_test.go#L79C2-L80C25

We should add this?

Reply:
I'm not sure we need to.

In the consuming with checkpoints example, the code creates a brand new context before starting the processor:

```go
processorCtx, processorCancel := context.WithCancel(context.TODO())
defer processorCancel()

if err := processor.Run(processorCtx); err != nil {
          // ...
}
```

So, using `defer processorCancel()` is the right call here.

Instead, the input v2 receives the `ctx` from the caller, so I guess the caller is probably responsible for canceling the context.

WDYT?

Reply:
> Instead, the input v2 receives the ctx from the caller, so I guess the caller is probably responsible for canceling the context.

Yeah, ideally yes.